### DISCUSSION TITLE
Scenario Goal revamp by splitting sub-goals

### DISCUSSION THREAD

**Discussion by MaxMallon** [OpenRCT2/OpenRCT2#16457] [Feature requests]
I think scenario goals are in need for improvement. Rather than a list like now, I think it'd be cool to see a number of checkmarks to combine different goal flags.
For example, rather than just "x number people, with rating 600, at end of year y", OR "X number of people without dropping rating below 700"
Have all the different subgoals be sepperately selectable:

Number of people:  0-60000
Park Value: 0-999999999
Due Date: none, 0-999
Minimal Park Rating: 0-999 
Park Rating must be maintained: Yes, No
Monthly Ride Ticket Sales: 0-1000000
Monthly Shop Income: 0-1000000
Number of coasters: 0-255
Minimal Length of Coasters: 0-10000
Minimal Excitement of Coasters: 0-10
Loan must be repayed: Yes, No

Basically have it be possible for some crazy difficult scenario where you gotta have like 4000 guests at the park at the end of year 4, never having the park rating drop below 900, having 10 coasters of at least 9 excitement, with the loan repayed, and a shop income of 20000 or more in one month. Stuff like that. 

Of course, simply adding more subgoals than what makes up the current selection would be great too, think of things like

Build x number of waterrides, with or without a certain excitement
Make x money from photo section sales
Receive x number of positive awards
Fail scenario upon receiving a negative award
Fail scenario upon going into the red (negative money)

Also perhaps it would be nice to have early-scenario completion be selectable as intended or not with an additional such flag. Some scenarios get harder as the time goes by, and thus get easier with early scenario completion on. It could be worded as "Achieve goal before the end of year 5", vs "Achieve the goal AT the end of year 5".

**Comment by ocalhoun6** [comment]
Ideally, the UI for this would be hidden behind an 'advanced options' button or something similar when creating a new scenario, to avoid overcomplicating things for scenario builders who might not be as experienced in it.

**Reply by MaxMallon** [reply]
![scenario](https://user-images.githubusercontent.com/22072563/149902392-d4576a1b-675f-438a-a82f-d2e3ee6b08df.png)
I was thinking something like this (photoshopped)

**Reply by duncanspumpkin** [reply]
I would suggest a tick box on the left hand side that decides if a goal is in use.

**Reply by IntelOrca** [reply]
A separate tab for the objective too, keep the name, and details on its own tab.

**Reply by ZehMatt** [reply]
Can't we make this a bit more flexible in general? What comes to mind looking at all this is basically a list of operations:
```(value) (operator) (target)```
So it would be neat if the user could select for value lets say "Guests", for the operator "Less than" and the target could be a value or maybe another variable, perhaps optionally negate the result via checkbox so logic can be inverse.

**Reply by MaxMallon** [reply]
> A separate tab for the objective too, keep the name, and details on its own tab.

Sepperate tab, or sepperate window? If it's in a tab, wouldn't players accidentally skip past it, not realizing this one window had extra tabs?



> Can't we make this a bit more flexible in general? What comes to mind looking at all this is basically a list of operations: `(value) (operator) (target)` So it would be neat if the user could select for value lets say "Guests", for the operator "Less than" and the target could be a value or maybe another variable, perhaps optionally negate the result via checkbox so logic can be inverse.

You mean like having a dropdown thingy on the left, where you select whether you're talking about guests/parkvalue/coasters etc, and then a -+ box next to it for the amount? It seems a bit tricky to me, as you'd have to build in logic to prevent them from selecting the same goal-type twice. And what would be the point of having "less than" goals? "Have less than 5000 guests in the park at the end of year 4?" Perhaps for coasters it could be interesting, win the scenario with at most 5 coasters, but other than that it seems a bit pointless

**Reply by ZehMatt** [reply]
That was purely an example to what an "operator" might be, the user would have a lot of freedom on deciding what the winning and loosing conditions are.

**Reply by ocalhoun6** [reply]
> Can't we make this a bit more flexible in general? What comes to mind looking at all this is basically a list of operations: `(value) (operator) (target)` So it would be neat if the user could select for value lets say "Guests", for the operator "Less than" and the target could be a value or maybe another variable, perhaps optionally negate the result via checkbox so logic can be inverse.

Ooh, that would be cool. Especially with creative use of 'less than' operators.

Imagine a scenario where you were required to get a monthly ride income of $x _while having less than 1000 guests in the park_. For a scenario that requires you to squeeze as much money as possible from a small number of guests.

Or 'must have 4000 guests in the park by year 4' + 'park rating must remain _below_ 500 at all times' for a scenario that requires you to attract a lot of guests to a park that *doesn't* make the guests happy.

Things like that could add really interesting new twists to scenario play. And you're right -- a lot of flexibility and potential for creativity could be built into the system with a `(value) (operator) (target)` type system for specifying scenario goals.

**Reply by MaxMallon** [reply]
I'm already working on it :) 
Including the less than bits

**Reply by fathertimejr** [reply]
Hey, what happened with this? I found my way here because I had an idea for a scenario involving the inverse of a normal scenario goal, so I was excited to see ZehMatt's suggestion. Adding more options, separating park rating from guest count, etc. sounds great. Mid-year deadlines would be nice, but I could maybe get the same effect another way.

Marcel commented something like: it's hard to make a difficult scenario; they usually end up just being tedious. I say new options for objectives could work wonders for the custom scenario scene and give the game's popularity a surge.

Btw... The limit currently goes beyond $1 million. I know because I held down the + button for a few minutes. Would be nice to be able to type in the number, but not necessary I guess.
>Monthly Ride Ticket Sales: 0-1000000

**Reply by MaxMallon** [reply]
Well, they didn't like what I made after I'd finished everything including UI, and I don't want to remake it entirely from scratch. So it's gone into the trash

**Comment by James103** [comment]
Additional variables that could be used as objective types after the base system is implemented:

Financial variables:
* Loan Amount
* Bank Balance
* Company Value
* Ride Construction Expenses
* Ride Running Costs
* Land Purchase Expenses
* Shop Stock Expenses
* Food/Drink Stock Expenses
* Staff Wages
* Marketing Expenses
* Research Expenses
* Loan Interest

Park statistic variables:
* Number of Staff
  - Number of Handymen
  - Number of Mechanics
  - Number of Security Guards
  - Number of Entertainers
* Number of Rides (Total)
* Number of Shops and Stalls
* Total Admissions

Example Objectives and Restrictions:
* Never take out a loan of more than £30,000
  - This value may be less than the true max loan, in which case a player may increase their loan above the target value but run the risk of having their park closed down.
* Have a bank balance of at most £10,000
* Have a company value of at least £250,000
* Never spend more than £5,000 on ride construction in one month
* Spend more than £1,000 on ride running costs in one month
* Spend at most £500 on staff wages in one month
* Never spend more than £200 on loan interest in one month
* Never spend less than £100 on marketing in one month
* Have at least 20 staff in your park
* Never have more than 5 handymen in your park
* Have at least 500 guests visit your park in total
  - Replicates the "sell at least X park tickets" goal from Parkitect.
